Make new modules importable

saraedum opened the pull request:
Fixes #5.

saraedum committed:
Make new modules importable

nthiery commented on recursive_monkey_patch.py:
```diff
@@ -206,6 +208,14 @@ def monkey_patch(source, target, log_level=logging.WARNING, logger=None):
         >>> A.NestedNew.__module__
         'a_test_module_patch.submodule'
+
+    .. RUBRIC:: Testing that new modules behave like old modules
```
Maybe: "like preexisting modules in the target"?

nthiery commented on recursive_monkey_patch.py:
```diff
@@ -206,6 +208,14 @@ def monkey_patch(source, target, log_level=logging.WARNING, logger=None):
+    The new modules can be imported (actually they are already imported)::
+
+        >>> import a_test_module.submodule_new
+        >>> from a_test_module.submodule_new import i
```
What about displaying the value of `a_test_module.submodule.__module__`?

saraedum replied:
Hm, `AttributeError: 'module' object has no attribute '__module__'`

nthiery replied:
Sorry, I meant `.__name__`

saraedum committed:
Rename comment

saraedum committed:
Merge remote-tracking branch 'origin/master' into import

nthiery commented:
I haven't yet wrapped around my head to understand and double check the new logic.
Is it sill possible to import the module under it's original name?
Could this possibly break (or actually improve) things like pickling of instances of classes defined in the module?

saraedum commented:
I did this to make pickling with factories work :)

saraedum commented:
Let me try the original import and add a doctest for it.

saraedum committed:
Document issues (?) relating to double importing

saraedum committed:
Fix double import issues

It is unclear what we want here. Importing the same module is twice is probably worse than having broken names.

nthiery commented:
:-)

Could the module name renaming confuse introspection? E.g. what happens upon
```
    sage: a_test_module.submodule_new??
    sage: a_test_module_patch.submodule_new??
```
and similar in classes/functions defined in `submodule_new`?

saraedum commented:
My last commit contains an alternative that is probably better?

saraedum commented:
Let me check introspection now.

saraedum commented:
Here is the behavior with my latest changes:
```
sage: from recursive_monkey_patch import monkey_patch
sage: import a_test_module_patch
sage: import a_test_module
sage: monkey_patch(a_test_module_patch, a_test_module)
sage: a_test_module.submodule_new??
Type:            module
String form:     <module 'a_test_module.submodule_new' from 'a_test_module_patch/submodule_new.py'>
File:            ~/proj/sage/recursive-monkey-patch/a_test_module_patch/submodule_new.py
Source:         
class New:
    r"""
    A docstring
    """
    pass
Class docstring:
module(name[, doc])

Create a module object. The name must be a string; the optional doc
argument can have any type.
Init docstring:  x.__init__(...) initializes x; see help(type(x)) for signature
sage: from a_test_module.submodule_new import New
sage: New??
Source:
class New:
    r"""
    A docstring
    """
    pass
File:   ~/proj/sage/recursive-monkey-patch/a_test_module_patch/submodule_new.py
Type:   classobj
sage: import a_test_module_patch.submodule_new
sage: a_test_module_patch.submodule_new??
Type:            module
String form:     <module 'a_test_module.submodule_new' from 'a_test_module_patch/submodule_new.py'>
File:            ~/proj/sage/recursive-monkey-patch/a_test_module_patch/submodule_new.py
Source:         
class New:
    r"""
    A docstring
    """
    pass
Class docstring:
module(name[, doc])

Create a module object. The name must be a string; the optional doc
argument can have any type.
Init docstring:  x.__init__(...) initializes x; see help(type(x)) for signature
sage: from a_test_module_patch.submodule_new import New as New_
sage: New_??
Source:
class New:
    r"""
    A docstring
    """
    pass
File:   ~/proj/sage/recursive-monkey-patch/a_test_module_patch/submodule_new.py
Type:   classobj
```

saraedum commented:
Looks reasonable to me.

saraedum committed:
Merge remote-tracking branch 'origin/master' into import

saraedum commented:
~~@nthiery Have you enabled travis for Pull Requests? Would be nice to see it checking this now :)~~

Great. Seems to work.

nthiery commented:
I agree the behaviour sounds sane. I did not get the time to review the logic; but it's not yet critical software, so tests + putting it into production will be good enough. Let's get this in.
Thanks much for the contribution!
